crates/gitbutler-branch-actions/src/commit_ops.rs

```diff
@@ -0,0 +1,238 @@
+        common_base.tree_id()?,
+        exclusive_superset,
+        exclusive_subset,
+        Default::default(),
+        options,
+    )?;
+
+    if merged_exclusives.conflicts.is_empty() {
+        Ok(exclusive_superset == merged_exclusives.tree.write()?)
+    } else {
+        Ok(false)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use gitbutler_testsupport::testing_repository::TestingRepository;
+    mod get_exclusive_tree {
+        use gitbutler_oxidize::OidExt;
+
+        use super::super::get_exclusive_tree;
+        use super::*;
+
+        #[test]
+        fn when_already_exclusive_returns_self() {
+            let test_repository = TestingRepository::open();
+            let base_commit: git2::Commit =
+                test_repository.commit_tree(None, &[("foo.txt", "foo"), ("bar.txt", "bar")]);
+            let second_commit: git2::Commit = test_repository.commit_tree(
+                Some(&base_commit),
+                &[("foo.txt", "bar"), ("bar.txt", "baz")],
+            );
+
+            let exclusive_tree = get_exclusive_tree(
+                &test_repository.gix_repository(),
+                second_commit.id().to_gix(),
+                base_commit.id().to_gix(),
+            )
+            .unwrap();
+
+            assert_eq!(
+                second_commit.tree_id().to_gix(),
+                exclusive_tree,
+                "The tree returned should match the second commit"
+            )
+        }
+
+        #[test]
+        fn when_on_top_of_other_commit_its_changes_are_dropped() {
+            let test_repository = TestingRepository::open();
+            let base_commit: git2::Commit =
+                test_repository.commit_tree(None, &[("foo.txt", "foo")]);
+            let second_commit: git2::Commit = test_repository.commit_tree(
+                Some(&base_commit),
+                &[("foo.txt", "bar"), ("bar.txt", "baz")],
+            );
+            let third_commit: git2::Commit = test_repository.commit_tree(
+                Some(&second_commit),
+                &[("foo.txt", "bar"), ("bar.txt", "baz"), ("qux.txt", "bax")],
+            );
+
+            // The second commit changed foo.txt, and added bar.txt. We expect
+            // foo.txt to be reverted back to foo, and bar.txt should be dropped
+            let expected_commit: git2::Commit =
+                test_repository.commit_tree(None, &[("foo.txt", "foo"), ("qux.txt", "bax")]);
+
+            let exclusive_tree = get_exclusive_tree(
+                &test_repository.gix_repository(),
+                third_commit.id().to_gix(),
+                base_commit.id().to_gix(),
+            )
+            .unwrap();
+
+            assert_eq!(expected_commit.tree_id().to_gix(), exclusive_tree,)
+        }
+    }
+
+    mod is_subset {
+        use gitbutler_oxidize::OidExt;
+
+        use super::super::is_subset;
+        use super::*;
+
+        #[test]
+        fn a_commit_is_a_subset_of_itself() {
```

Comment:
All these tests are read-only, which is when the `read_only::fixture*` methods are faster. They create a fixture only once and make it easy to inspect on disk. They come at the disadvantage of separating the test-setup from the test itself though, but to me it's easy enough to check the script and I can understand more easily how the structure looks like by looking at it, and if in doubt the final structure is easy to inspect on disk.

Of course, these are not the pinnacle of what's possible and I always thought that it would be great to define tests more like `cargo` does, but to apply the same techniques to easily define complex Git repositories.